activerecord/lib/active_record/connection_adapters/sqlite3/database_statements.rb

```diff
@@ -100,7 +104,11 @@ def perform_query(raw_connection, sql, binds, type_casted_binds, prepare:, notif
                   stmt.step
                   ActiveRecord::Result.empty
                 else
-                  ActiveRecord::Result.new(stmt.columns, stmt.to_a)
+                  types = {}
+                  stmt.columns.zip(stmt.types).each_with_index do |(c, t), i|
+                    types[c] = types[i] = type_map.lookup(t)
+                  end
+                  ActiveRecord::Result.new(stmt.columns, stmt.to_a, types.freeze)
```

Comment:
I get that you copied what was done for PG, but this implementation isn't great performance wise. I probably never noticed it because I never profile with pg.

I'll experiment a few things.